Update "Extending Tripal" to mention plugins and services plus add an overview for upgrading existing modules to Tripal 4 (#96)
tripal/tripal_doc: docs/dev_guide/module_dev/plugin_service.rst

Plugins and Services (Building blocks)
=================================================

What is a plugin?
------------------

You can think of a plugin as defining a type of functionality (e.g. Field,
Importer) that (a) has or will have multiple implementations and (b) a single

> site may want to use 1+ of these implementations. A plugin is a sort of template or description that can be

Reviewer: ```suggestion
site may want to use 1 or more of these implementations. A plugin is a sort of template or description that can be
```

Plugins and Services (Building blocks)
=================================================

What is a plugin?
------------------

You can think of a plugin as defining a type of functionality (e.g. Field,

> Importer) that has or will have multiple implementations for which a single

Reviewer: ```suggestion
Importer) that has or will have multiple implementations which a single
```

> site may want to use. A plugin is a sort of template or description that can be

Reviewer 2: ```suggestion
Importer) that (a) has or will have multiple implementations and (b) a single
site may want to use 1+ of these implementations. A plugin is a sort of template or description that can be
```

used in multiple different situations, rather than explicitly coding a single
field or a single importer that is only used in one situation.

For example, if your module imports data into a Tripal site, you are *most
likely* going to create an implementation of the Tripal Importer class rather
than define your own plugin. A field is an example of a plugin created by
someone else that you would want to implement. In order for your module to
provide a custom Tripal field, then you are going to be implementing either the
:doc:`Tripal Field<tripal_fields>` plugin or Chado Field plugin.

Although more complex than typical plugins, entities are also technically an

> example of a plugin. Read more about :doc:`Entities<entities>` to help decide

Reviewer 2: I'm not 100% sure that entities are actually plugins. They should be but I suspect the were developed independantly at the same time despite full-filling a similar niche. What about this wording instead because I agree that we want people to look at all 3 as options!

```suggestion
Although more complex than typical plugins, entities are fill a similar role in modules. Read more about :doc:`Entities<entities>` to help decide
```
